File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/bulk/BulkResponseItem.java
package fr.myprysm.vertx.elasticsearch.action.bulk;

import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;
import org.elasticsearch.action.DocWriteRequest;
import org.elasticsearch.rest.RestStatus;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@DataObject(generateConverter = true)
public class BulkResponseItem {

    private DocWriteRequest.OpType opType;
    private RestStatus status;
    private String index;
    private String type;
    private String id;
    private Long version;
    private boolean failed;
    private String failure;

    public BulkResponseItem(BulkResponseItem other) {
        opType = other.opType;
        status = other.status;
        index = other.index;
        type = other.type;
        id = other.id;
        version = other.version;
        failed = other.failed;
        failure = other.failure;
    }

    /**
     * Build a new <code>BulkResponseItem</code> from a <code>JsonObject</code>.
     *
     * @param json the <code>JsonObject</code>
     */
    public BulkResponseItem(JsonObject json) {
        BulkResponseItemConverter.fromJson(json, this);
    }

    /**
     * Transforms the <code>BulkResponseItem</code> into a <code>JsonObject</code>.
     *
     * @return the <code>JsonObject</code>
     */
    public JsonObject toJson() {
        JsonObject json = new JsonObject();
        BulkResponseItemConverter.toJson(this, json);
        return json;
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/bulk/BulkResponseItem.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.bulk

import fr.myprysm.vertx.elasticsearch.action.bulk.BulkResponseItem
import org.elasticsearch.action.DocWriteRequest.OpType
import org.elasticsearch.rest.RestStatus

fun BulkResponseItem(
  failed: Boolean? = null,
  failure: String? = null,
  id: String? = null,
  index: String? = null,
  opType: OpType? = null,
  status: RestStatus? = null,
  type: String? = null,
  version: Long? = null): BulkResponseItem = fr.myprysm.vertx.elasticsearch.action.bulk.BulkResponseItem().apply {

  if (failed != null) {
    this.setFailed(failed)
  }
  if (failure != null) {
    this.setFailure(failure)
  }
  if (id != null) {
    this.setId(id)
  }
  if (index != null) {
    this.setIndex(index)
  }
  if (opType != null) {
    this.setOpType(opType)
  }
  if (status != null) {
    this.setStatus(status)
  }
  if (type != null) {
    this.setType(type)
  }
  if (version != null) {
    this.setVersion(version)
  }
}

